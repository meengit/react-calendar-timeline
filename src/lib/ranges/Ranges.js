import React, { Component, PropTypes } from 'react'
import Range from './Range'
import { _get, arraysEqual } from '../utils'

export default class Ranges extends Component {
  static propTypes = {
    canvasTimeEnd: PropTypes.number.isRequired,
    canvasTimeStart: PropTypes.number.isRequired,
    canvasWidth: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    headerHeight: PropTypes.number.isRequired,
    keys: PropTypes.object.isRequired,
    ranges: PropTypes.array.isRequired,
    visibleTimeStart: PropTypes.number.isRequired,
    visibleTimeEnd: PropTypes.number.isRequired
  }

  shouldComponentUpdate (nextProps, nextState) {
    return !(nextProps.canvasTimeStart === this.props.canvasTimeStart &&
             nextProps.canvasTimeEnd === this.props.canvasTimeEnd &&
             nextProps.canvasWidth === this.props.canvasWidth &&
             nextProps.height === this.props.height &&
             nextProps.keys === this.props.keys &&
             nextProps.headerHeight === this.props.headerHeight &&
             arraysEqual(nextProps.ranges, this.props.ranges) &&
             nextProps.visibleTimeStart === this.props.visibleTimeStart &&
             nextProps.visibleTimeEnd === this.props.visibleTimeEnd
      )
  }

  getVisibleRanges (visibleTimeStart, visibleTimeEnd, ranges) {
    return this.props.ranges.filter(range => {
      let position = {
        start: range.start,
        end: range.end
      }

      if (visibleTimeStart >= range.start) {
        position.start = visibleTimeStart
      }

      if (visibleTimeEnd <= range.end) {
        position.end = visibleTimeEnd
      }

      if (visibleTimeStart > position.end || visibleTimeEnd < position.start) {
        position = {}
        return
      }

      return position
    })
  }

  rangeIsVisible (range) {
    function isNumber (value) {
      return value !== undefined && typeof (value) === 'number' && !isNaN(value)
    }

    if (isNumber(range.start) && isNumber(range.end)) {
      return range
    }
  }

  render () {
    const { visibleTimeStart, visibleTimeEnd } = this.props
    const { rangeIdKey } = this.props.keys
    let visibleRanges = this.getVisibleRanges(visibleTimeStart, visibleTimeEnd, this.props.ranges)

    return (
      <div className='rct-ranges'>
        {visibleRanges.filter(range => this.rangeIsVisible(range))
                      .map(range => <Range canvasTimeStart={this.props.canvasTimeStart}
                                           canvasTimeEnd={this.props.canvasTimeEnd}
                                           canvasWidth={this.props.canvasWidth}
                                           className={range.className}
                                           height={this.props.height}
                                           headerHeight={this.props.headerHeight}
                                           key={_get(range, rangeIdKey)}
                                           keys={this.props.keys}
                                           range={range}
                                           rangeStart={range.start}
                                           rangeEnd={range.end} />)}
      </div>
    )
  }
}
