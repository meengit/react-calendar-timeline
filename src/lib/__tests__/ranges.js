import React from 'react'
import { render } from 'enzyme'
import Ranges from '../ranges/Ranges'

const defaults = {
  canvasWidth: 3000,
  height: 1190,
  headerHeight: 60,
  visibleTimeStart: 3000,
  visibleTimeEnd: 6000
}

const zoom = defaults.visibleTimeEnd - defaults.visibleTimeStart // 3000
const canvasTimeStart = defaults.visibleTimeStart - (defaults.visibleTimeEnd - defaults.visibleTimeStart) // 0
const canvasTimeEnd = canvasTimeStart + zoom * 3 // 9000

const keys = {
  rangeIdKey: 'id',
  rangeTimeStartKey: 'start',
  rangeTimeEndKey: 'end'
}

describe('Ranges', () => {
  it('render Range in visible time', () => {
    const wrapper = render(
      <Ranges canvasTimeStart={canvasTimeStart}
              canvasTimeEnd={canvasTimeEnd}
              canvasWidth={defaults.canvasWidth}
              height={defaults.height}
              headerHeight={defaults.headerHeight}
              keys={keys}
              ranges={[{id: 1, start: 4000, end: 5000}]}
              visibleTimeStart={defaults.visibleTimeStart}
              visibleTimeEnd={defaults.visibleTimeEnd}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('do not render Range with end smaller than visible time start', () => {
    const wrapper = render(
      <Ranges canvasTimeStart={canvasTimeStart}
              canvasTimeEnd={canvasTimeEnd}
              canvasWidth={defaults.canvasWidth}
              height={defaults.height}
              headerHeight={defaults.headerHeight}
              keys={keys}
              ranges={[{id: 2, start: 1000, end: 2000}]}
              visibleTimeStart={defaults.visibleTimeStart}
              visibleTimeEnd={defaults.visibleTimeEnd}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('do not render Range with start after than visible time end', () => {
    const wrapper = render(
      <Ranges canvasTimeStart={canvasTimeStart}
              canvasTimeEnd={canvasTimeEnd}
              canvasWidth={defaults.canvasWidth}
              height={defaults.height}
              headerHeight={defaults.headerHeight}
              keys={keys}
              ranges={[{id: 3, start: 7000, end: 8000}]}
              visibleTimeStart={defaults.visibleTimeStart}
              visibleTimeEnd={defaults.visibleTimeEnd}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('render Range that overlaps visible time start', () => {
    const wrapper = render(
      <Ranges canvasTimeStart={canvasTimeStart}
              canvasTimeEnd={canvasTimeEnd}
              canvasWidth={defaults.canvasWidth}
              height={defaults.height}
              headerHeight={defaults.headerHeight}
              keys={keys}
              ranges={[{id: 4, start: 2000, end: 4000}]}
              visibleTimeStart={defaults.visibleTimeStart}
              visibleTimeEnd={defaults.visibleTimeEnd}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('render Range that overlaps visible time end', () => {
    const wrapper = render(
      <Ranges canvasTimeStart={canvasTimeStart}
              canvasTimeEnd={canvasTimeEnd}
              canvasWidth={defaults.canvasWidth}
              height={defaults.height}
              headerHeight={defaults.headerHeight}
              keys={keys}
              ranges={[{id: 5, start: 5000, end: 7000}]}
              visibleTimeStart={defaults.visibleTimeStart}
              visibleTimeEnd={defaults.visibleTimeEnd}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
