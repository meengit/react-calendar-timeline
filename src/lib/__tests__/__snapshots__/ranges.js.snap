// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ranges do not render Range with end smaller than visible time start 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "rct-ranges",
        },
        "children": Array [],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;

exports[`Ranges do not render Range with start after than visible time end 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "rct-ranges",
        },
        "children": Array [],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;

exports[`Ranges render Range in visible time 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "rct-ranges",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "rct-range",
              "style": "top:60px;left:1333px;height:1130px;width:334px;",
            },
            "children": Array [],
            "name": "div",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
          },
        ],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;

exports[`Ranges render Range that overlaps visible time end 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "rct-ranges",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "rct-range",
              "style": "top:60px;left:1667px;height:1130px;width:666px;",
            },
            "children": Array [],
            "name": "div",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
          },
        ],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;

exports[`Ranges render Range that overlaps visible time start 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "rct-ranges",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "rct-range",
              "style": "top:60px;left:667px;height:1130px;width:666px;",
            },
            "children": Array [],
            "name": "div",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
          },
        ],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;
